igvc_navigation/src/trajectory_planner/smooth_control.cpp

```diff
@@ -0,0 +1,211 @@
+}
+
+RobotState SmoothControl::acquireNewTarget(const nav_msgs::PathConstPtr& path, const RobotState& state,
+                                           size_t state_index) const
+{
+  geometry_msgs::Point end = path->poses[path->poses.size() - 1].pose.position;
+
+  // target position
+  RobotState newTarget{ end.x, end.y, tf::getYaw(path->poses[path->poses.size() - 1].pose.orientation) };
+
+  double distance = 0;
+
+  for (size_t i = state_index; i < path->poses.size() - 1; i++)
+  {
+    geometry_msgs::Point point1, point2;
+    point1 = path->poses[i].pose.position;
+    point2 = path->poses[i + 1].pose.position;
+    double increment = igvc::get_distance(point1.x, point1.y, point2.x, point2.y);
+
+    if (distance + increment > target_selection_options_.lookahead_dist)
+    {
+      newTarget.set_x(point2.x);
+      newTarget.set_y(point2.y);
+      newTarget.set_yaw(tf::getYaw(path->poses[i + 1].pose.orientation));
+      return newTarget;
+    }
+    distance += increment;
+  }
+
+  return newTarget;
+}
+
+size_t SmoothControl::findTargetOnPath(const nav_msgs::PathConstPtr& path, const RobotState& target) const
+{
+  auto closest = std::min_element(path->poses.begin(), path->poses.end(),
+                                  [&target](geometry_msgs::PoseStamped lhs, geometry_msgs::PoseStamped rhs) {
+                                    return target.distTo(lhs.pose.position) < target.distTo(rhs.pose.position);
+                                  });
+  if (target.distTo(closest->pose.position) < target_selection_options_.target_move_threshold)
+  {
+    return static_cast<size_t>(closest - path->poses.begin());
+  }
+  ROS_WARN_THROTTLE(1, "Target moved more than target_move_threshold");
+  return 0;
+}
+
+double SmoothControl::distAlongPath(const nav_msgs::PathConstPtr& path, size_t path_index, size_t target_index) const
+{
+  double distance = -1;
+
+  for (size_t i = path_index; i < target_index; i++)
+  {
+    geometry_msgs::Point point1, point2;
+    point1 = path->poses[i].pose.position;
+    point2 = path->poses[i + 1].pose.position;
+    distance += igvc::get_distance(point1.x, point1.y, point2.x, point2.y);
+  }
+  return distance;
+}
+
+bool SmoothControl::reachedTarget(const RobotState& state, const RobotState& target) const
+{
+  return state.distTo(target) < target_selection_options_.target_reached_distance;
+}
+
+size_t SmoothControl::getClosestIndex(const nav_msgs::PathConstPtr& path, const RobotState& state) const
+{
+  auto closest = std::min_element(path->poses.begin(), path->poses.end(),
+                                  [&state](geometry_msgs::PoseStamped lhs, geometry_msgs::PoseStamped rhs) {
+                                    return state.distTo(lhs.pose.position) < state.distTo(rhs.pose.position);
+                                  });
+  return static_cast<size_t>(closest - path->poses.begin());
+}
+
+double SmoothControl::blendCurvature(double distance, double K1, double K2) const
```

Comment:
This really shouldn't be necessary with a small enough dt, just as a heads up.

Reply:
it's mentioned in the paper, but ehh you're probably right.